### PR TITLE
Scripting: get context names REST API

### PR DESCRIPTION
**Pull request by stu-elastic**
Scripting: get context names REST API

Adds `GET /_script_context`, returning a `contexts` object with each
available context as a key whose value is an empty object. eg.
```
{
  "contexts": {
    "aggregation_selector": {},
    "aggs": {},
    "aggs_combine": {},
...
  }
}
```

refs: #47411

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
Looking good! I have a few comments.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 1]
```diff
@@ -538,6 +539,10 @@ public StoredScriptSource getStoredScript(ClusterState state, GetStoredScriptReq
         }
     }
 
+    public List<String> getContextNames() {
+        return new ArrayList<>(contexts.keySet());
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Agreed, will change to `Set`.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    public static GetScriptContextResponse fromXContent(XContentParser parser) throws IOException {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        GetScriptContextResponse that = (GetScriptContextResponse) o;
+        return new HashSet<>(contextNames).equals(new HashSet<>(that.contextNames));
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java]: :+1:

File: server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        );
+
+    static {
+        PARSER.declareNamedObjects(
+            ConstructingObjectParser.constructorArg(),
+            (p, c, n) ->
+            {
+                // advance empty object
+                p.nextToken();
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java]: Done.

File: server/src/test/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        if (randomBoolean()) {
+            return new GetScriptContextResponse(Collections.emptyMap());
+        }
+        Map<String,Object> items = new HashMap<>();
+        for (int i = randomIntBetween(1, 10); i > 0; i--) {
+            items.put(randomAsciiLettersOfLengthBetween(1, 16), new Object());
+        }
+        return new GetScriptContextResponse(items);
+
+    }
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/test/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponseTests.java]: Done.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetScriptContextResponse.java]: Done.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/TransportGetScriptContextAction.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class TransportGetScriptContextAction extends TransportMasterNodeReadAction<GetScriptContextRequest, GetScriptContextResponse> {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/TransportGetScriptContextAction.java]: Nope, changed to `HandledTransportAction`.

File: server/src/main/java/org/elasticsearch/client/ClusterAdminClient.java [hunk 6]
```diff
@@ -718,4 +720,9 @@
      * Get a script from the cluster state
      */
     ActionFuture<GetStoredScriptResponse> getStoredScript(GetStoredScriptRequest request);
+
+    /**
+     * Get a script from the cluster state
+     */
+    void getScriptContext(GetScriptContextRequest request, ActionListener<GetScriptContextResponse> listener);
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/client/ClusterAdminClient.java]: Removed.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Changed.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by stu-elastic** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by colings86** [comment]
@stu-elastic Please remember to label all PRs with change type, area and version labels. This is important for the release notes process as well as letting people looking at the PR know quickly where it is targetting

**Comment by colings86** [comment]
I've added some labels I think are appropriate but please check and amend as necessary

**Pull request merged**